include/ginkgo/core/solver/batch_solver_base.hpp

```diff
@@ -0,0 +1,297 @@
+     */
+    double get_residual_tolerance() const { return residual_tol_; }
+
+    /**
+     * Update the residual tolerance to be used by the solver.
+     *
+     * @param res_tol  The residual tolerance to be used for subsequent
+     *                 invocations of the solver.
+     */
+    void set_residual_tolerance(double res_tol) { residual_tol_ = res_tol; }
+
+    /**
+     * Get the maximum number of iterations set on the solver.
+     *
+     * @return  Maximum number of iterations.
+     */
+    int get_max_iterations() const { return max_iterations_; }
+
+    /**
+     * Set the maximum number of iterations for the solver to use,
+     * independent of the factory that created it.
+     *
+     * @param max_iterations  The maximum number of iterations for the solver.
+     */
+    void set_max_iterations(int max_iterations)
+    {
+        max_iterations_ = max_iterations;
+    }
+
+protected:
+    BatchSolver() {}
+
+    BatchSolver(std::shared_ptr<const BatchLinOp> system_matrix,
+                std::shared_ptr<const BatchLinOp> gen_preconditioner,
+                const double res_tol, const int max_iterations)
+        : system_matrix_{std::move(system_matrix)},
+          preconditioner_{std::move(gen_preconditioner)},
+          residual_tol_{res_tol},
+          max_iterations_{max_iterations}
+    {}
+
+    std::shared_ptr<const BatchLinOp> system_matrix_{};
+    std::shared_ptr<const BatchLinOp> preconditioner_{};
+    double residual_tol_{};
+    int max_iterations_{};
+};
+
+
+namespace detail {
+
+
+struct common_batch_params {
+    std::shared_ptr<const BatchLinOpFactory> prec_factory;
+    std::shared_ptr<const BatchLinOp> generated_prec;
+    double residual_tolerance;
+    int max_iterations;
+};
+
+
+template <typename ParamsType>
+common_batch_params extract_common_batch_params(ParamsType& params)
+{
+    return {params.preconditioner, params.generated_preconditioner,
+            params.default_residual_tol, params.default_max_iterations};
+}
+
+
+}  // namespace detail
+
+
+struct BatchInfo {
+    std::unique_ptr<log::BatchLogDataBase> logdata;
+};
+
+
+/**
+ * @tparam PolymorphicBase  The base class; must be a subclass of BatchLinOp.
+ */
+template <typename ConcreteSolver,
+          typename ValueType = typename ConcreteSolver::value_type,
+          typename PolymorphicBase = BatchLinOp>
+class EnableBatchSolver
+    : public BatchSolver,
+      public EnableBatchLinOp<ConcreteSolver, PolymorphicBase> {
+public:
+    const ConcreteSolver* apply(ptr_param<const MultiVector<ValueType>> b,
+                                ptr_param<MultiVector<ValueType>> x) const
+    {
+        this->validate_application_parameters(b.get(), x.get());
+        auto exec = this->get_executor();
+        this->apply_impl(make_temporary_clone(exec, b).get(),
+                         make_temporary_clone(exec, x).get());
+        return self();
+    }
+
+    const ConcreteSolver* apply_impl(
```

Comment:
why is this public?

Reply:
Also, why do we have two apply_impl?

Reply:
this was an error, should not be there now